app/model/RuleContext.scala

```diff
@@ -47,24 +47,28 @@ case class RuleContext(authContext: AuthContext)(implicit hc: HeaderCarrier) {
     enrolmentSeq.filter(_.state != EnrolmentState.ACTIVATED).map(_.key).toSet[String]
   }
 
-  lazy val lastSaReturn = authContext.principal.accounts.sa
-    .fold(Future(SaReturn.empty))(saAccount => selfAssessmentConnector.lastReturn(saAccount.utr.value))
+  lazy val lastSaReturn = authority.flatMap { auth => auth.saUtr.fold(Future(SaReturn.empty))(saUtr => selfAssessmentConnector.lastReturn(saUtr.utr)) }
 
-  lazy val currentCoAFEAuthority = frontendAuthConnector.currentCoAFEAuthority()
+  lazy val authority = credId match {
+    case Some(credId) => authConnector.tarAuthority(credId)
```

Comment:
name shadowing. you will need a different name in the `Some(xyz)` or maybe you can use `_`